web/src/search_suggestion.ts

```diff
@@ -842,20 +842,67 @@ function suggestion_search_string(suggestion_line: SuggestionLine): string {
     return search_strings.join(" ");
 }
 
+function suggestions_for_current_filter(): SuggestionLine[] {
+    if (narrow_state.stream_name() && narrow_state.topic() !== "") {
+        return [
+            get_default_suggestion_line([
+                {
+                    operator: "channel",
+                    operand: narrow_state.stream_name()!,
+                },
+            ]),
+            get_default_suggestion_line(narrow_state.search_terms()),
+        ];
+    }
+    if (narrow_state.pm_emails_string()) {
+        return [
+            get_default_suggestion_line([
+                {
+                    operator: "is",
+                    operand: "dm",
+                },
+            ]),
+            get_default_suggestion_line(narrow_state.search_terms()),
+        ];
+    }
+    return [get_default_suggestion_line(narrow_state.search_terms())];
+}
+
 class Attacher {
     result: SuggestionLine[] = [];
     prev = new Set<string>();
     base: SuggestionLine;
+    add_current_filter: boolean;
 
-    constructor(base: SuggestionLine) {
+    constructor(base: SuggestionLine, search_query_is_empty: boolean, add_current_filter: boolean) {
         this.base = base;
+        this.add_current_filter = add_current_filter;
+        // Sometimes we add suggestions with the current filter in case
+        // the user wants to search within the current filter. For an empty
+        // search query, we put the current filter suggestions at the start
+        // of the list.
+        if (search_query_is_empty && this.add_current_filter) {
+            this.add_current_filter = false;
+            for (const current_filter_line of suggestions_for_current_filter()) {
+                this.push(current_filter_line);
+            }
+        }
     }
 
     push(suggestion_line: SuggestionLine): void {
         const search_string = suggestion_search_string(suggestion_line);
         if (!this.prev.has(search_string)) {
             this.prev.add(search_string);
             this.result.push(suggestion_line);
+            // Sometimes we add suggestions with the current filter in case
+            // the user wants to search within the current filter. If we haven't
+            // added this yet, do it after the first suggestion.
+            if (this.add_current_filter) {
+                this.add_current_filter = false;
+                for (const current_filter_line of suggestions_for_current_filter()) {
+                    this.push([...current_filter_line, ...suggestion_line]);
+                }
+            }
```

Comment:
Dropping this hunk because the PR is useful without it, and doing so avoids the nasty repetition bug. I spent 30m and couldn't find a pattern I liked for how best to add this "second option" with existing filters; something like `check_validity` in theory would be useful, but it seems like we'd need to do some refactoring to have these `SuggestionLine` objects contain a new attribute with the set of `NarrowTerm` objects that they represent, or something like that.

```
diff --git a/web/src/search_suggestion.ts b/web/src/search_suggestion.ts
index 4f5943e246..118df6e8cf 100644
--- a/web/src/search_suggestion.ts
+++ b/web/src/search_suggestion.ts
@@ -894,15 +894,6 @@ class Attacher {
         if (!this.prev.has(search_string.toLowerCase())) {
             this.prev.add(search_string.toLowerCase());
             this.result.push(suggestion_line);
-            // Sometimes we add suggestions with the current filter in case
-            // the user wants to search within the current filter. If we haven't
-            // added this yet, do it after the first suggestion.
-            if (this.add_current_filter) {
-                this.add_current_filter = false;
-                for (const current_filter_line of suggestions_for_current_filter()) {
-                    this.push([...current_filter_line, ...suggestion_line]);
-                }
-            }
         }
     }
 
```